hazelcast/src/main/java/com/hazelcast/osgi/impl/HazelcastOSGiServiceImpl.java

```diff
@@ -0,0 +1,387 @@
+        if (!active) {
+            throw new IllegalStateException("Hazelcast OSGI Service is not active!");
+        }
+    }
+
+    private Config getConfig(Config config) {
+        if (config == null) {
+            config = new XmlConfigBuilder().build();
+        }
+        GroupConfig groupConfig = config.getGroupConfig();
+        if (groupConfig == null
+                || StringUtil.isNullOrEmpty(groupConfig.getName())
+                || GroupConfig.DEFAULT_GROUP_NAME.equals(groupConfig.getName())) {
+            if (!Boolean.getBoolean(HAZELCAST_OSGI_GROUPING_DISABLED)) {
+                String groupName = id;
+                if (groupConfig == null) {
+                    config.setGroupConfig(new GroupConfig(groupName));
+                } else {
+                    groupConfig.setName(groupName);
+                }
+            }
+        }
+        return config;
+    }
+
+    private HazelcastInstance createHazelcastInstance(Config config) {
+        return Hazelcast.newHazelcastInstance(getConfig(config));
+    }
+
+    private HazelcastOSGiInstance registerInstance(HazelcastInstance instance) {
+        HazelcastOSGiInstance hazelcastOSGiInstance;
+        if (instance instanceof HazelcastOSGiInstance) {
+            hazelcastOSGiInstance = (HazelcastOSGiInstance) instance;
+        } else {
+            hazelcastOSGiInstance = new HazelcastOSGiInstanceImpl(instance, this);
+        }
+        ServiceRegistration serviceRegistration;
+        if (!Boolean.getBoolean(HAZELCAST_OSGI_REGISTER_DISABLED)) {
+            serviceRegistration =
+                    ownerBundleContext.registerService(HazelcastInstance.class.getName(), hazelcastOSGiInstance, null);
+        } else {
+            serviceRegistration = EMPTY_SERVICE_REGISTRATION;
+        }
+        instanceServiceRegistrationMap.put(hazelcastOSGiInstance, serviceRegistration);
+        instanceMap.put(instance.getName(), hazelcastOSGiInstance);
+        return hazelcastOSGiInstance;
+    }
+
+    private void deregisterInstance(HazelcastOSGiInstance hazelcastOSGiInstance) {
+        instanceMap.remove(hazelcastOSGiInstance.getName());
+        ServiceRegistration serviceRegistration =
+                instanceServiceRegistrationMap.remove(hazelcastOSGiInstance);
+        if (serviceRegistration != null && serviceRegistration != EMPTY_SERVICE_REGISTRATION) {
+            ownerBundleContext.ungetService(serviceRegistration.getReference());
+        }
+    }
+
+    private void shutdownAllInternal() {
+        try {
+            for (Map.Entry<HazelcastOSGiInstance, ServiceRegistration> entry : instanceServiceRegistrationMap.entrySet()) {
+                HazelcastOSGiInstance instance = entry.getKey();
+                ServiceRegistration serviceRegistration = entry.getValue();
+                if (serviceRegistration != EMPTY_SERVICE_REGISTRATION) {
+                    try {
+                        ownerBundleContext.ungetService(serviceRegistration.getReference());
```

Comment:
Could call deregisterInstance here first before shutting down (so it's consistent with shutdownInstance)